skcosmo/sparse/IncrementalSparseKRR.py

```diff
@@ -0,0 +1,147 @@
+#!/usr/bin/env python
+
+import os
+import sys
+import numpy as np
+from tools import sorted_eigh
+
+class IncrementalSparseKRR(object):
+    """
+        Performs sparsified kernel ridge regression. Example usage:
+
+        KMM = build_kernel(Xm, Xm)
+        iskrr = IncrementalSparseKRR()
+        iskrr.initialize_fit(KMM)
+        for i in batches:
```

Comment:
What do you think about hiding the batches in the fit function as sklearn it does for incremental pca?
https://github.com/scikit-learn/scikit-learn/blob/0fb307bf3/sklearn/decomposition/_incremental_pca.py#L210-L211
Is there use case where it is useful for you to do it yourself in a script?

Reply:
The way I have the current fitting procedure set up is more conceptually similar to the sklearn incremental PCA `partial_fit`. When working with kernels that are really big and you don't want to store them in memory but can compute them in batches, I find it useful to explicitly loop over the batches where in each iteration the "partial kernel" is computed and directly passed to `fit_batch`.

You raise a good point though, we could also have a `fit` function that does the loop internally. The trick here would then be constructing infrastructure to handle both structure-based and environment-based kernels.

Reply:
Can't your input for this class be already structural features/environmental features? Then you construct the kernel as in sklearn.kernel_ridge.KernelRidge. Or am I missing something?

https://scikit-learn.org/stable/modules/generated/sklearn.kernel_ridge.KernelRidge.html#sklearn.kernel_ridge.KernelRidge

Reply:
This would certainly work for environment features. For structural features, things are more complicated depending on whether you are using a linear or nonlinear kernel. For structure-based kernel methods, you get per-atom properties through the mean of the environment kernel values averaged over each structure. Some references for this are the supplementary info of [this article](https://doi.org/10.1126/sciadv.1701816) and Section IIB and IIIF [here](https://arxiv.org/abs/1901.10971).

For a linear kernel, you can take a shortcut: compute the average feature vector for each structure, and compute the kernel between the averages.

For a nonlinear kernel, you have to compute the kernel between individual environments and then average over the values in the kernel matrix. It's more difficult to develop a notion of "structural features" in a nonlinear kernel setting.

To make things as general and clean as possible, here I just take the kernel(s) as input. I think as we move forward and begin to add kernel methods to the main repo, this is a development discussion we should have, i.e., whether the user is required to build their own kernel beforehand, or we have internal machinery that computes the structure averaging over the environment kernel. The latter is probably more sklearn-like, and it shouldn't be too painful. In #32 I have a wrapper for kernel functions that computes this kind of averaging.